beacon_node/beacon_chain/src/canonical_head.rs

```diff
@@ -0,0 +1,1307 @@
+
+        // Sanity check the finalized checkpoint.
+        //
+        // The new finalized checkpoint must be either equal to or better than the previous
+        // finalized checkpoint.
+        check_against_finality_reversion(&old_view, &new_view)?;
+
+        let new_head_proto_block = fork_choice_read_lock
+            .get_block(&new_view.head_block_root)
+            .ok_or(Error::HeadBlockMissingFromForkChoice(
+                new_view.head_block_root,
+            ))?;
+
+        // Do not allow an invalid block to become the head.
+        //
+        // This check avoids the following infinite loop:
+        //
+        // 1. A new block is set as the head.
+        // 2. The EL is updated with the new head, and returns INVALID.
+        // 3. We call `process_invalid_execution_payload` and it calls this function.
+        // 4. This function elects an invalid block as the head.
+        // 5. GOTO 2
+        //
+        // In theory, fork choice should never select an invalid head (i.e., step #3 is impossible).
+        // However, this check is cheap.
+        if new_head_proto_block.execution_status.is_invalid() {
+            return Err(Error::HeadHasInvalidPayload {
+                block_root: new_head_proto_block.root,
+                execution_status: new_head_proto_block.execution_status,
+            });
+        }
+
+        // Exit early if the head or justified/finalized checkpoints have not changed, there's
+        // nothing to do.
+        if new_view == old_view {
+            debug!(
+                self.log,
+                "No change in canonical head";
+                "head" => ?new_view.head_block_root
+            );
+            return Ok(None);
+        }
+
+        // Get the parameters to update the execution layer since either the head or some finality
+        // parameters have changed.
+        let new_forkchoice_update_parameters =
+            fork_choice_read_lock.get_forkchoice_update_parameters();
+
+        perform_debug_logging::<T>(&old_view, &new_view, &fork_choice_read_lock, &self.log);
+
+        // Drop the read lock, it's no longer required and holding it any longer than necessary
+        // will just cause lock contention.
+        drop(fork_choice_read_lock);
+
+        // If the head has changed, update `self.canonical_head`.
+        let new_cached_head = if new_view.head_block_root != old_view.head_block_root {
+            metrics::inc_counter(&metrics::FORK_CHOICE_CHANGED_HEAD);
+
+            // Try and obtain the snapshot for `beacon_block_root` from the snapshot cache, falling
+            // back to a database read if that fails.
+            let new_snapshot = self
+                .snapshot_cache
+                .try_read_for(BLOCK_PROCESSING_CACHE_LOCK_TIMEOUT)
+                .and_then(|snapshot_cache| {
+                    snapshot_cache.get_cloned(
+                        new_view.head_block_root,
+                        CloneConfig::committee_caches_only(),
```

Comment:
Is it necessary to clone with the committee caches given that you end up calling `snapshot.beacon_state.build_all_committee_caches(&self.spec)` below?